src/cargo/core/compiler/context/compilation_files.rs

```diff
@@ -144,11 +144,13 @@ impl<'a, 'cfg: 'a> CompilationFiles<'a, 'cfg> {
     /// target.
     pub fn out_dir(&self, unit: &Unit<'a>) -> PathBuf {
         if unit.mode.is_doc() {
-            self.layout(unit.kind).root().parent().unwrap().join("doc")
+            self.layout(unit.kind).doc().to_path_buf()
         } else if unit.target.is_custom_build() {
             self.build_script_dir(unit)
         } else if unit.target.is_example() {
             self.layout(unit.kind).examples().to_path_buf()
+        } else if unit.mode == CompileMode::Test || unit.mode == CompileMode::Bench {
+            self.layout(unit.kind).legacy_deps().to_path_buf()
```

Comment:
I think that tests & benches should be treated as "public" outputs: it's not uncommon to launch test executable manually, to attach gdb. Should we use the following layout?

```
target/release
  examples/
  tests/
  benches/
  bins & libs
```

?